pkg/ingester/index/multi.go

```diff
@@ -0,0 +1,123 @@
+package index
+
+import (
+	"time"
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/querier/astmapper"
+	"github.com/grafana/loki/pkg/storage/config"
+)
+
+type periodIndex struct {
+	time.Time
+	idx int // address of the index to use
+}
+
+type Multi struct {
+	periods []periodIndex
+	indices []Interface
+}
+
+func NewMultiInvertedIndex(periods []config.PeriodConfig, indexShards uint32) (*Multi, error) {
+	var (
+		err error
+
+		ii          Interface // always stored in 0th index
+		bitPrefixed Interface // always stored in 1st index
+
+		periodIndices []periodIndex
+	)
+
+	for _, pd := range periods {
+		switch pd.IndexType {
+		case config.TSDBType:
+			if bitPrefixed == nil {
+				bitPrefixed, err = NewBitPrefixWithShards(indexShards)
+				if err != nil {
+					return nil, err
+				}
+			}
+			periodIndices = append(periodIndices, periodIndex{
+				Time: pd.From.Time.Time(),
+				idx:  1, // tsdb inverted index is always stored in position one
+			})
+		default:
+			if ii == nil {
+				ii = NewWithShards(indexShards)
+			}
+			periodIndices = append(periodIndices, periodIndex{
+				Time: pd.From.Time.Time(),
+				idx:  0, // regular inverted index is always stored in position zero
+			})
+		}
+	}
+
+	return &Multi{
+		periods: periodIndices,
+		indices: []Interface{ii, bitPrefixed},
+	}, nil
+}
+
+func (m *Multi) Add(labels []logproto.LabelAdapter, fp model.Fingerprint) (result labels.Labels) {
+	for _, i := range m.indices {
+		if i != nil {
+			result = i.Add(labels, fp)
+		}
+	}
+	return
+}
+
+func (m *Multi) Delete(labels labels.Labels, fp model.Fingerprint) {
+	for _, i := range m.indices {
+		if i != nil {
+			i.Delete(labels, fp)
+		}
+	}
+
+}
+
+func (m *Multi) Lookup(t time.Time, matchers []*labels.Matcher, shard *astmapper.ShardAnnotation) ([]model.Fingerprint, error) {
+	return m.indexFor(t).Lookup(matchers, shard)
+}
+
+func (m *Multi) LabelNames(t time.Time, shard *astmapper.ShardAnnotation) ([]string, error) {
+	return m.indexFor(t).LabelNames(shard)
+}
+
+func (m *Multi) LabelValues(t time.Time, name string, shard *astmapper.ShardAnnotation) ([]string, error) {
+	return m.indexFor(t).LabelValues(name, shard)
+}
+
+// Query planning is responsible for ensuring no query spans more than one inverted index.
```

Comment:
is this comment out of place?

Reply:
I intended to put it here because most queries in loki use a `from, through`, but in this case we're using another part of the code base to ensure we'd never query more than one period at a time, so we don't need to account for the range `from->through`, only the first point (`from`).